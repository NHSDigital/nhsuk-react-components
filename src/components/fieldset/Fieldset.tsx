import FieldsetContext, { IFieldsetContext } from './FieldsetContext';
import HeadingLevel, { HeadingLevelType } from '../../util/HeadingLevel';
import React, { HTMLProps, PureComponent } from 'react';

import { NHSUKSize } from '../../util/types/NHSUKTypes';
import classNames from 'classnames';

interface LegendProps extends Omit<HTMLProps<HTMLLegendElement>, 'size'> {
  isPageHeading?: boolean;
  headingLevel?: HeadingLevelType;
  size?: NHSUKSize;
}

const Legend: React.FC<LegendProps> = ({
  className,
  children,
  isPageHeading,
  headingLevel,
  size,
  ...rest
}) => (
  <legend
    className={classNames(
      'nhsuk-fieldset__legend',
      {
        'nhsuk-fieldset__legend--xl': isPageHeading && !size,
      },
      { [`nhsuk-fieldset__legend--${size}`]: size },
      className,
    )}
    {...rest}
  >
    {isPageHeading ? (
      <HeadingLevel className="nhsuk-fieldset__heading" headingLevel={headingLevel}>
        {children}
      </HeadingLevel>
    ) : (
      children
    )}
  </legend>
);

Legend.defaultProps = {
  headingLevel: 'h1',
};


interface FieldsetProps extends HTMLProps<HTMLFieldSetElement> {
  disableErrorLine?: boolean;
}

type FieldsetState = { registeredComponents: Array<string>; erroredComponents: Array<string> };

class Fieldset extends PureComponent<FieldsetProps, FieldsetState> {
  static Legend = Legend;

  constructor(props: FieldsetProps) {
    super(props);
    this.state = {
      registeredComponents: [],
      erroredComponents: [],
    };
  }

  passError = (componentId: string, error: boolean) => {
    this.setState(state => {
      const existingError = state.erroredComponents.includes(componentId);
      if (existingError && !error) {
        return {
          ...state,
          erroredComponents: state.erroredComponents.filter(id => id !== componentId),
        };
      }
      if (!existingError && error) {
        return { ...state, erroredComponents: [...state.erroredComponents, componentId] };
      }
      return state;
    });
  };

  registerComponent = (componentId: string, deregister: boolean = false) => {
    this.setState(state => {
      if (deregister) {
        return {
          ...state,
          registeredComponents: state.registeredComponents.filter(id => id !== componentId),
        };
      }
      if (!state.registeredComponents.includes(componentId)) {
        return {
          ...state,
          registeredComponents: [...state.registeredComponents, componentId],
        };
      }
      return state;
    });
  };


  
  render() {
    const { className, disableErrorLine, ...rest } = this.props;
    const contextValue: IFieldsetContext = {
      isFieldset: true,
      registerComponent: this.registerComponent,
      passError: this.passError,
    };

    const containsFormElements = this.state.registeredComponents.length > 0;
    const containsError = this.state.erroredComponents.length > 0;

    return (
      <FieldsetContext.Provider value={contextValue}>
        {containsFormElements ? (
          <div
            className={classNames('nhsuk-form-group', {
              'nhsuk-form-group--error': disableErrorLine ? false : containsError,
            })}
          >
            <fieldset className={classNames('nhsuk-fieldset', className)} {...rest} />
          </div>
        ) : (
          <fieldset className={classNames('nhsuk-fieldset', className)} {...rest} />
        )}
      </FieldsetContext.Provider>
    );
  }
}

Fieldset.Legend = Legend;

export default Fieldset;
