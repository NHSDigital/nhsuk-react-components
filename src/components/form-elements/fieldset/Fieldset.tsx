import React, { FC, HTMLProps, MutableRefObject, useMemo, useState } from 'react';
import classNames from 'classnames';
import { NHSUKSize } from '@util/types/NHSUKTypes';
import HeadingLevel, { HeadingLevelType } from '@util/HeadingLevel';
import FieldsetContext, { IFieldsetContext } from './FieldsetContext';

interface LegendProps extends Omit<HTMLProps<HTMLLegendElement>, 'size'> {
  isPageHeading?: boolean;
  headingLevel?: HeadingLevelType;
  size?: NHSUKSize;
}

const Legend: FC<LegendProps> = ({
  className,
  children,
  isPageHeading,
  headingLevel = 'h1',
  size,
  ...rest
}) => (
  <legend
    className={classNames(
      'nhsuk-fieldset__legend',
      {
        'nhsuk-fieldset__legend--xl': isPageHeading && !size,
      },
      { [`nhsuk-fieldset__legend--${size}`]: size },
      className,
    )}
    {...rest}
  >
    {isPageHeading ? (
      <HeadingLevel className="nhsuk-fieldset__heading" headingLevel={headingLevel}>
        {children}
      </HeadingLevel>
    ) : (
      children
    )}
  </legend>
);

interface FieldsetProps extends HTMLProps<HTMLFieldSetElement> {
  fieldsetRef?: MutableRefObject<HTMLFieldSetElement | null>;
  disableErrorLine?: boolean;
}

const FieldSet = ({ className, disableErrorLine, fieldsetRef, ...rest }: FieldsetProps) => {
  const [registeredComponents, setRegisteredComponents] = useState<string[]>([]);
  const [erroredComponents, setErroredComponents] = useState<string[]>([]);

  const passError = (componentId: string, error: boolean): void => {
    const existingError = erroredComponents.includes(componentId);
    if (existingError && !error) {
      setErroredComponents(erroredComponents.filter((id) => id !== componentId));
      return;
    }
    if (!existingError && error) {
      setErroredComponents([...erroredComponents, componentId]);
    }
  };

  const registerComponent = (componentId: string, deregister = false): void => {
    let newComponents = [...registeredComponents];
    if (deregister) {
      newComponents = newComponents.filter((id) => id !== componentId);
    } else if (!registeredComponents.includes(componentId)) {
      newComponents = [...newComponents, componentId];
    }
    setRegisteredComponents(newComponents);
  };

  const contextValue: IFieldsetContext = useMemo(() => {
    return {
      isFieldset: true,
      registerComponent: registerComponent,
      passError: passError,
    };
  }, [registerComponent, passError]);

  const containsFormElements = registeredComponents.length > 0;
  const containsError = erroredComponents.length > 0;

  return (
    <FieldsetContext.Provider value={contextValue}>
      {containsFormElements ? (
        <div
          className={classNames('nhsuk-form-group', {
            'nhsuk-form-group--error': disableErrorLine ? false : containsError,
          })}
        >
          <fieldset
            className={classNames('nhsuk-fieldset', className)}
            ref={fieldsetRef}
            {...rest}
          />
        </div>
      ) : (
        <fieldset className={classNames('nhsuk-fieldset', className)} ref={fieldsetRef} {...rest} />
      )}
    </FieldsetContext.Provider>
  );
};

FieldSet.Legend = Legend;

export default FieldSet;
