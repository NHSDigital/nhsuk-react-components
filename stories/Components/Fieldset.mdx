import {Meta} from "@storybook/blocks";

<Meta title="Components/Fieldset" />

# Fieldset

This component can be found in the `nhsuk-frontend` repository <a href="https://github.com/nhsuk/nhsuk-frontend/tree/master/packages/components/fieldset" target="_blank" rel="noopener noreferrer">here</a>.

## Implementation Notes

The `Fieldset` component has one subcomponent: `Fieldset.Legend`.

The `Fieldset.Legend` component has one default prop: `headingLevel: 'h1'`. This can be overridden and any valid heading level can be used (i.e. `h1`, `h2`).

## Usage

### Standard

```jsx
import { Fieldset } from "nhsuk-react-components";

const Element = () => {
    return (
        <Fieldset>
            <Fieldset.Legend>What is your address?</Fieldset.Legend>
        </Fieldset>
    );
}
```
