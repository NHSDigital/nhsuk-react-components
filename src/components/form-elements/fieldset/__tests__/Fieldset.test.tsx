import React from 'react';
import { shallow } from 'enzyme';
import Fieldset from '..';

describe('Fieldset', () => {
  it('matches snapshot', () => {
    const element = shallow(<Fieldset>Text</Fieldset>);
    expect(element.text()).toBe('Text');
    expect(element).toMatchSnapshot('Fieldset');
    element.unmount();
  });

  describe('Fieldset.Legend', () => {
    it('matches snapshot', () => {
      const element = shallow(<Fieldset.Legend>Text</Fieldset.Legend>);
      expect(element).toMatchSnapshot('FieldsetLegend');
      element.unmount();
    });

    it('renders as page heading', () => {
      const element = shallow(<Fieldset.Legend isPageHeading>Text</Fieldset.Legend>);
      expect(element.hasClass('nhsuk-fieldset__legend--xl')).toBeTruthy();
      expect(
        element
          .find('.nhsuk-fieldset__heading')
          .render()
          .text(),
      ).toBe('Text');
      element.unmount();
    });
  });
});
