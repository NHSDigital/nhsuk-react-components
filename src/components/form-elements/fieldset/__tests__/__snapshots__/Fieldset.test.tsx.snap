// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fieldset Fieldset.Legend matches snapshot: FieldsetLegend 1`] = `
<legend
  className="nhsuk-fieldset__legend"
>
  Text
</legend>
`;

exports[`Fieldset matches snapshot: Fieldset 1`] = `
<ContextProvider
  value={
    Object {
      "isFieldset": true,
      "passError": [Function],
      "registerComponent": [Function],
    }
  }
>
  <fieldset
    className="nhsuk-fieldset"
  >
    Text
  </fieldset>
</ContextProvider>
`;
